app/src/main/java/com/example/marvelapp/database/MarvelDataBaseImpl.kt

```diff
@@ -0,0 +1,37 @@
+package com.example.marvelapp.database
+
+import androidx.room.Database
+import androidx.room.RoomDatabase
+import com.example.marvelapp.api.util.Result
+import com.example.marvelapp.database.entity.CharacterEntity
+import com.example.marvelapp.database.mapper.CharacterDataBaseMapper
+import com.example.marvelapp.entity.Character
+
+@Database(
+    entities = [CharacterEntity::class],
+    version = 1
+)
+abstract class MarvelDataBaseImpl : RoomDatabase(), MarvelDataBase {
+
+    abstract fun marvelDao(): MarvelDao
+    private val characterMapper = CharacterDataBaseMapper()
+
+    override fun getAllCharacters(): Result<List<Character>> =
+        marvelDao().getAllCharacters().let {
+            if (it.isNotEmpty()) {
+                Result.Success(characterMapper.transformToListOfCharacter(it))
+            } else {
+                Result.Failure(Exception(CHARACTER_ERROR))
+            }
+        }
+
+    override fun updateCharacters(characters: List<Character>) {
+        characters.forEach {
+            marvelDao().insertCharacter(characterMapper.transformToData(it))
+        }
+    }
+
+    companion object {
+        private const val CHARACTER_ERROR = "Characters not found, please connect internet"
```

Comment:
if you will this line to show it on the screen should be in the string res, not here

Reply:
done